std/arith.asm

```diff
@@ -17,18 +23,176 @@ machine Arith(CLK32_31, operation_id){
+    pol commit x2__0, x2__1, x2__2, x2__3, x2__4, x2__5, x2__6, x2__7, x2__8, x2__9, x2__10, x2__11, x2__12, x2__13, x2__14, x2__15;
+    let x2 = [x2__0, x2__1, x2__2, x2__3, x2__4, x2__5, x2__6, x2__7, x2__8, x2__9, x2__10, x2__11, x2__12, x2__13, x2__14, x2__15];
+    pol commit y2__0, y2__1, y2__2, y2__3, y2__4, y2__5, y2__6, y2__7, y2__8, y2__9, y2__10, y2__11, y2__12, y2__13, y2__14, y2__15;
+    let y2 = [y2__0, y2__1, y2__2, y2__3, y2__4, y2__5, y2__6, y2__7, y2__8, y2__9, y2__10, y2__11, y2__12, y2__13, y2__14, y2__15];
+    pol commit x3__0, x3__1, x3__2, x3__3, x3__4, x3__5, x3__6, x3__7, x3__8, x3__9, x3__10, x3__11, x3__12, x3__13, x3__14, x3__15;
+    let x3 = [x3__0, x3__1, x3__2, x3__3, x3__4, x3__5, x3__6, x3__7, x3__8, x3__9, x3__10, x3__11, x3__12, x3__13, x3__14, x3__15];
+    pol commit y3__0, y3__1, y3__2, y3__3, y3__4, y3__5, y3__6, y3__7, y3__8, y3__9, y3__10, y3__11, y3__12, y3__13, y3__14, y3__15;
+    let y3 = [y3__0, y3__1, y3__2, y3__3, y3__4, y3__5, y3__6, y3__7, y3__8, y3__9, y3__10, y3__11, y3__12, y3__13, y3__14, y3__15];
+
+    let s_for_eq1 = |x1, y1, x2, y2, limb_index| (div(sub(y2, y1), sub(x2, x1)) >> (limb_index * 16)) & 0xffff;
+    let s_for_eq2 = |x1, y1, limb_index| (div(mul(3, mul(x1, x1)), mul(2, y1)) >> (limb_index * 16)) & 0xffff;
+
+    let compute_x3_int = |x1, x2, s| (s * s - x1 - x2 + 2 * secp_modulus) % secp_modulus;
+    let compute_y3_int = |x1, y1, x3, s| (s * (((x1 - x3) + secp_modulus) % secp_modulus) - y1 + secp_modulus) % secp_modulus;
+
+    // Note that the most significant limb is 32-Bit instead of 16
+    let compute_q0_for_eq1 = |x1, y1, x2, y2, s, limb_index| ((-(s * x2 - s * x1 - y2 + y1) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+    let compute_q0_for_eq2 = |x1, y1, s, limb_index| ((-(2 * s * y1 - 3 * x1 * x1) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+    let compute_q1 = |x1, x2, x3, s, limb_index| ((-(s * s - x1 - x2 - x3) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+    let compute_q2 = |x1, y1, x3, y3, s, limb_index| ((-(s * x1 - s * x3 - y1 - y3) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+ 
+    let limbs_to_int = |limbs, row| array::sum(array::new(array::len(limbs), |i| int(limbs[i](row)) << (i * 16)));
+
+    let x1_int: int -> int = |row| limbs_to_int(x1, row);
+    let y1_int: int -> int = |row| limbs_to_int(y1, row);
+    let x2_int: int -> int = |row| limbs_to_int(x2, row);
+    let y2_int: int -> int = |row| limbs_to_int(y2, row);
+    let x3_int: int -> int = |row| limbs_to_int(x3, row);
+    let s_int: int -> int = |row| limbs_to_int(s, row);
+
+    let s_hint = |i, limb_index| if selEq[1](i) == fe(1) {
```

Comment:
Ok, here we really need to evaluate a witness column. So for this I need to add a builtin `eval` function, but I think this should be doable.